### DISCUSSION TITLE
Error while performing iterative mid-circuit measurements

### DISCUSSION THREAD

**Discussion by Gopal-Dahale** [NVIDIA/cuda-quantum#113] [Q&A]
I am trying to perform mid-circuit measurements and apply unitary based on the outcome. Below is the sample code:

```cpp
// Compile and run with:

#include <cudaq.h>

struct kernel {
  void operator()(const int steps) __qpu__ {
    cudaq::qreg q(2);

    for (int i = 0; i < steps; ++i) {
      
      // Prepare the initial state by superposition
      h(q[1]);

      // Mid circuit measurement
      auto c = mz(q[1]);

      if(c){
        x(q[0]);
      }

      reset(q[1]);   
    }
  }
};

int main() {

  using namespace cudaq::spin;

  // Problem parameters
  const int nShots = 100;
  const int steps = 100;


  // Sample
  auto counts = cudaq::sample(nShots, kernel{}, steps);

  // Fine grain access to the bits and counts
  for (auto &[bits, count] : counts) {
    printf("Observed: %s, %lu\n", bits.data(), count);
  }

}
```
I was able to run the program with the following commands without any error.

```
nvq++ test.cpp
./a.out
```

For small value of `steps` (like 10) the following commands works fine too but for 100, it gives error on running `./a.out`.

```
nvq++ --qpu cuquantum test.cpp
./a.out
```
 
Error:

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  [custatevec] %internal error in sample (line 513)
Aborted (core dumped)
```
Any help in this regard is appreciated.

**Comment by amccaskey** [comment]
Why do you need the inner for loop? Would the following work for you? 
```cpp
#include <cudaq.h>

struct kernel {
  void operator()() __qpu__ {
    cudaq::qreg q(2);
      
    // Prepare the initial state by superposition
    h(q[1]);

    // Mid circuit measurement
    auto c = mz(q[1]);

    if(c){
      x(q[0]);
    }
  }
};

int main() {
  const int nShots = 100;
  auto counts = cudaq::sample(nShots, kernel{});
  counts.dump();
}
```
should output 
```bash
{ 
  __global__ : { 00:52 11:48 }
   c : { 0:52 1:48 }
}
```

Granted, that we should be throwing a better error message for the case when sampling with an inner for loop.

**Reply by Gopal-Dahale** [reply]
The inner loop is part of the algorithm that I want to use (that's why its iterative). Can you provide me with some way to continue with the code I shared?

**Reply by stevescia** [reply]
@Gopal-Dahale curious about your test, I compiled and ran your code; the program ran without an error (on cpu);  
(i'm using a gitpod.io ubuntu box: "Ubuntu 22.04.2 LTS")

$ /workspace/cuda-quantum/cuda-quantum/build/bin/nvq++ test.cpp -o a.out
$ ./a.out 
Observed: 10, 46
Observed: 00, 54
(with 100 steps)

If you're still crashing, try to build the target in debug with the -g flag:

$ /workspace/cuda-quantum/cuda-quantum/build/bin/nvq++ -v test.cpp -o debug.a.out -g 
(the -v flag is verbose mode shows the compiler steps i prefer to use)

load it into gdb and run it, seeing where it crashes..

$ gdb debug.a.out
(gdb) run

**Reply by Gopal-Dahale** [reply]
@stevescia Thank you for response. On CPU, the code works fine but on with the `--qpu cuquantum` flag, I got the error I mentioned earlier. I executed the code in debug mode with the flag and still got the error.

```
nvq++ -v flow.cpp --qpu cuquantum -o debug.a.out -g
gdb debug.a.out
```
Error:
```
Starting program: /home/cudaq/examples/debug.a.out 
warning: Error disabling address space randomization: Operation not permitted
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7f0433dff000 (LWP 5941)]
[New Thread 0x7f042d2d2000 (LWP 5964)]
[New Thread 0x7f042cad1000 (LWP 5965)]
[New Thread 0x7f0427e7c000 (LWP 5966)]
terminate called after throwing an instance of 'std::runtime_error'
  what():  [custatevec] %internal error in measureQubit (line 443)

Thread 1 "debug.a.out" received signal SIGABRT, Aborted.
0x00007f0460dd8a7c in pthread_kill () from /usr/lib/x86_64-linux-gnu/libc.so.6
```

**Reply by stevescia** [reply]
@Gopal-Dahale ah i see, i'll get the cuQuantum going and see if i can debug as well..  you should be able to use gdb comands to view stack trace at point of failure,  etc 

bt - Prints a stack trace
